source/grid/grid_out.cc

```diff
@@ -2526,6 +2529,107 @@ void GridOut::write_vtu (const Triangulation<dim,spacedim> &tria,
 
 
 
+template <int dim, int spacedim>
+void GridOut::write_debug_vtu_per_processor (const Triangulation<dim,spacedim> &tria,
+                                             std::string              filename,
+                                             bool                     view_levels,
+                                             bool                     include_artificial) const
+{
+  std::vector<DataOutBase::Patch<dim,spacedim> > patches;
+  unsigned int n_datasets=4;
+  std::vector<std::string> data_names;
+  data_names.push_back("subdomain");
+  data_names.push_back("lvlsubdomain");
+  data_names.push_back("level");
+  data_names.push_back("proc_writing");
+
+  unsigned int n_q_points=GeometryInfo<dim>::vertices_per_cell;
+
+  typename Triangulation<dim, spacedim>::cell_iterator cell, endc;
+  for (cell=tria.begin(), endc=tria.end();
+       cell != endc; ++cell)
+    {
+      if (!include_artificial && cell->level_subdomain_id() ==
+          numbers::artificial_subdomain_id)
+        continue;
+      if (!view_levels && cell->has_children())
+        continue;
+
+      DataOutBase::Patch<dim,spacedim> patch;
+      patch.data.reinit(n_datasets, n_q_points);
+      patch.points_are_available = false;
+
+      for (unsigned int vertex=0; vertex<n_q_points; ++vertex)
+        {
+          patch.vertices[vertex] = cell->vertex(vertex);
+          patch.data(0,vertex) = cell->level();
+          if (!cell->has_children())
+            patch.data(1,vertex) =
+              (double)static_cast<int>(cell->subdomain_id());
+          else
+            patch.data(1,vertex) = -1.0;
+          patch.data(2,vertex) =
+            (double)static_cast<int>(cell->level_subdomain_id());
+          patch.data(3,vertex) =
+            (double)static_cast<int>(Utilities::MPI::this_mpi_process(MPI_COMM_WORLD));
```

Comment:
we should use ``Triangulation::locally_owned_subdomain()`` instead. Also remove the static cast to int here. It is only necessary for the subdomain above to get "-1" (otherwise it would be something like 4 billion).